uint8 num_nodes
Node[] nodes